Convert to a single change. (#1)
atomicincrement/atomicincrement.github.io: _posts/2021-11-18-polynomial-approximation.md

---
layout: post

> title:  "Making faster stats functions."

Reviewer: Can we have a snapper and more attention grabbing title? I'll also try and think of one

date:   2021-11-18
categories: maths
---

## Doctor Syn

Recently we have been working on a Rust library called "Doctor Syn", a play

> on the excellent **syn** library in Rust and Russel Thorndyke's character

Reviewer: ```suggestion
on the excellent (syn)[https://docs.rs/syn/1.0.81/syn/] library in Rust and (Russel Thorndyke's character)[https://en.wikipedia.org/wiki/Doctor_Syn]
```

who is both a priest and smuggler on the Romney marshes.

Doctor Syn's primary focus at present is to generate accurate polynomial

> approximations to functions. You are probably famililar with

Reviewer: ```suggestion
approximations to key functions important to the execution of many programs. You are probably familiar with
```

many of the functions we are targeting:

| Rust Function | calculates |
|---------------|------------|
| f32/f64::sin      | $$\sin{x}$$ |
| f32/f64::cos      | $$\cos{x}$$|
| f32/f64::atan2      | $$\arctan{y/x}$$|
| f32/f64::exp      | $$e^x$$|
| f32/f64::ln      | $$\log{x}$$|

[See Wikipedia](https://en.wikipedia.org/wiki/C_mathematical_functions)

> But we are mostly focused on statistical functions such as:

Reviewer: ```suggestion
While improving these functions has a lot of value, we are currently focusing most of our effort on statistical functions such as:
```

| R Function | distribution | role | calculates |
|------------|--------------|------------|
| dnorm      | normal | pdf | $$\frac{1}{\sigma\sqrt{2\pi}} e^{-\frac{1}{2}\left(\frac{x - \mu}{\sigma}\right)^2}$$|
| pnorm      | normal | cdf | $$\frac{1}{2}\left[1 + \operatorname{erf}\left( \frac{x-\mu}{\sigma\sqrt{2}}\right)\right] $$|
| qnorm      | normal | quantile | $$\mu+\sigma\sqrt{2} \operatorname{erf}^{-1}(2p-1)$$|
| rnorm      | normal | random | $$ \operatorname{qnorm}(\operatorname{runif}(i)) $$|

[See the R documention](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/Normal.html)

These functions are used extensively in finance and bioinformatics to perform statistical
inference to discover the parameters of models. For example, rnorm is often used to calculate

> the next step in the Metropolis-Hastings algorithm or to simulate option pricing.

Reviewer: Expanding on use cases
```suggestion
These functions are used extensively in finance and bioinformatics to perform statistical
inference, stochastic modelling, AI and Machine learning. For example, rnorm is a key part of many
MCMC algorithms and variational techniques, as well as a key part of monte carlo simulations, such as those that may be used to solve stochastic differential equations.
```

Using this libary, combined with parallel iterators, we can generate more efficent versions of

* Numpy
* R

> * GNU Octave

Reviewer: What would be the best examples we could give here?

and many others.

> We can also target new architectures like Arm SVE which do not fit the X86 model.

Reviewer: It's a bit much to jump to this having not discussed architectures or how they are relevant at this point. Perhaps we could either discuss around this point a bit more here, or leave it for the autovectorisation section.

We are working with the Isambard A64FX cluster to attempt to improve existing

> algorithms.

Reviewer: We should probably say a bit more about how Isambard fits in, and/or what we mean by improving algorithms and how this fits in.

## Surely this is a solved problem?

Whilst there are both fast or accurate versions of many of these functions, none of
them are able utilise modern compiler technology such as autovectorisation and
thread based parallelism efficiently. Also there are no fast *and* accurate versions

> of these functions, so we usually have to pick one.

Reviewer: rewrite to link the two points more closely

```suggestion
While great effort has been expended in key function optimization, current techniques are unable to efficiently utilize   modern compiler technology such as auto-vectorisation and thread based parallelism. This leads to the situation in which users are forced to trade off accuracy and speed, with no implementation that is both *fast* and *accurate*.
```

[See this for an example of *accurate*](https://blog.sigplan.org/2021/08/26/high-performance-correctly-rounded-math-libraries-for-32-bit-floating-point-representations/)

[See this discussion on *fast* on Stack Overflow](https://stackoverflow.com/questions/18662261/fastest-implementation-of-sine-cosine-and-square-root-in-c-doesnt-need-to-b)

If you really need to be accurate, it is best not to use floating point at all or to use
a larger floating point type than the one you are calculating. For example for **float16** use
**float32** to do the calculation and for **float32** use **float64**. For **float64** you will
be better off using fixed point integer arithmetic. Solutions that use table lookup are very
poor performers on 2020-era CPUs, but may get better as the **gather** instructions become

> optimised in SIMD implementations.

Reviewer: This remains an excellent point, but I'm a bit unclear on how the discussion on using higher precision floats supports the rest of this section. Could we tie this more explicitly into the discussion on performance vs accuracy, or  perhaps move it somewhere else?

Existing functions will not vectorise primarily because:

* They are in shared or static libraries.
* They contain branches and look-up tables.

These things were not a problem in the 1970's when many of these functions
were written, but today this is a problem. So we can't just inline the
legacy functions and hope that they will vectorise because they are

> not designed to do so.

Reviewer: Tied points together a bit more, tied back to central message more

```suggestion
In the 1970's when many of these functions were written, this was the state of the art. Today however, this is a problem: with the advances of modern vectorising processors, these implementations are substantially less efficient than they could be.

There is no short and simple fix to this problem either, these problems are fundamental ones that preclude any efficient vectorisation. The only viable approach to this problem is with substantial and novel changes such as the one that Doctor Syn proposes.
```

## Autovectorisation

In the past we would write our fast functions in assembler or even write assembler
to write our functions in machine code. But we like to think that we are more civilised
now. Such functions were hard to read and impossible to improve. Many such functions
hang around like a bad smell and more are being produced by chip vendors for special

> architectures. Excellent libraries like [Sleef](https://github.com/shibatch/sleef/blob/master/src/arch/helperavx512f.h) are done this way with machine specific intrinsics .

Reviewer: Tidied up slightly

```suggestion
In the past we have implemented fast functions in assembler or even written assembler
to write functions in machine code. These days however, we often try to take a more "civilized" approach where possible. Assembler functions are fast, but difficult to read, difficult to improve, and difficult to generalize. Despite these problems, many such functions end up hanging around like a bad smell, and more are being produced by chip vendors for special architectures. Even excellent libraries like [Sleef](https://github.com/shibatch/sleef/blob/master/src/arch/helperavx512f.h) are done this way with machine specific intrinsics.
```

But to make our code portable, for example to run on the new ARM SVE with variable sized registers,
we need a way of writing regular C or Rust code and having the compiler convert our loops into
SIMD code. We try to avoid writing SIMD code ourselves, these days, instead we make sure that

> our code will vectorised on modern compilers.

Reviewer: Tidied up, fitted into rewrite of last paragraph better

```suggestion
To solve the problems with these architecture specific assembly implementations, we try and write portable code; we wish our code to be able to run on both x86 architectures with SIMD as well as the new ARM SVE with variable sized registers. The way we achieve this is by trying to write code in such a way as that it will be automatically vectorised by modern compilers.
```

```rust
fn inc_doubles_scalar(x: &mut [f64]) {
    for x in x {
        *x += 1;
    }
}
```

[Godbolt](https://godbolt.org/z/qG5asYMGn)

This is much easier to read, works on all known hardware without modifications and
does not specify a vector size, which might be variable.

Vectorisers are fickle beasts however. If the wind blows in the wrong direction, the compiler
will often fail to vectorise or worse - vectorise in the IR and then convert the vector
operations to a long series of library calls.

For example, the following rather innocent function, which absolutely should be vectorisable,
converts itself into a series of function calls:

```rust
pub fn vector_sin(d: &mut [f64]) {
    d.iter_mut().for_each(|d| *d = f64::sin(*d))
}
```

[Goldbolt](https://godbolt.org/z/x4jr6MdM3) gives:

```
.LBB0_6:
        vmovsd  xmm0, qword ptr [r15 + 8*r12]
        vmovsd  qword ptr [rsp], xmm0
        vmovsd  xmm0, qword ptr [r15 + 8*r12 + 8]
        call    r14
        ... and 7 more calls like this.
        vunpcklpd       xmm0, xmm0, xmmword ptr [rsp]
        vmovups xmmword ptr [r15 + 8*r12 + 48], xmm0
        add     r12, 8
        add     rbp, 4
        jne     .LBB0_6
```

> Each call will take several hundred cycles.

Reviewer: I remain strongly of the opinion that while it's a good to discuss this problem, but we also need to discuss and explicitly link in a solution.

## Making library functions that vectorise

> Library functions make things bad for themselves by introducing
> branching.
> So even if we can inline the function, the functions will not vectorise.
> To get better accuracy, they divide the domain
> of a function - for example $$[-\pi, \pi]$$ for $$\sin(x)$$ into many small
> parts. This is often done using a `switch` statement which will not vectorise.

Reviewer: Tidy up

```suggestion
Existing library functions make vectorising a problem for themselves by introducing
branching. If a library function includes branching, the function will not vectorise even if it is inlined.
A common example of this is the division of the domain of the function into parts - for example dividing $$[-\pi, \pi]$$ for the $$\sin(x)$$ into a range of smaller domains, and providing a different implementation for each. This is often done using a branching `switch` statement which, as we have said,  will not vectorise.
```

Alternatives include using a lookup table of coefficients but many CPUs have not yet
implemented an efficient `gather` operation which can do table lookups
in reasonable time. The exception to this is GPUs, which do commonly
have efficient `gather` but these are likely to hurt the cache performance

> unless you use non-temporal loads and stores.

Reviewer: I think this would benefit from being more explicitly linked to it's context in the vectorisation discussion

So a good maths function will often sacrifice a little accuracy - one or two bits
to simplify the execution, but this is a choice that must be made. In the humble
opininion of the author, if you want more accuracy, use a larger number type
such a f64 for f32 and i64 (fixed point) for f64 calculations. This enables
you to calculate functions like sin and cos as a single polynomial instead of dividing

> it into sections.

Reviewer: This doesn't connect very well into the previous paragraph(s) of this section.

Author: Perhaps we should ditch it entirely.

Reviewer: Sounds reasonable to me

**Doctor Syn** generates functions that are free of complex control flow
which would inhibit vectorisation. The functions are all available as source code

> allowing inlining and as a result the chance of them inlining is much increased.

Reviewer: The second sentence is a little difficult for me to parse

    let x = scaled;
    let recip: fty = 1.0 / (x * x - 0.5 * 0.5);
    let y: fty = (from_bits(4730221388428958202u64))
        .mul_add(x * x, -from_bits(4731626383781768040u64))
        .mul_add(x * x, from_bits(4727627778628654481u64))
        .mul_add(x * x, -from_bits(4720012863723153492u64))
        .mul_add(x * x, from_bits(4708869911609092829u64))
        .mul_add(x * x, -from_bits(4695087533321972728u64))
        .mul_add(x * x, from_bits(4678670384600451257u64))
        .mul_add(x * x, -from_bits(4658680898319303328u64))
        .mul_add(x * x, from_bits(4635605149421499302u64))
        .mul_add(x * x, from_bits(4578476110820645018u64))
        .mul_add(x * x, from_bits(4611041379213747643u64))
        .mul_add(x * x, -from_bits(4603819697584151827u64))
        * x;
    y * recip
}
```

|Library|Function|ns per iteration (smaller is better)|
|-------|----|----------------|
|Doctor Syn|rnorm|2.4|
|Doctor Syn|parallel rnorm|0.9|
|rand_distr|Normal::sample()|6.9|
|rand_distr|parallel Normal::sample()|1.7|
|R|rnorm|60.0|
|Numpy|numpy.random.uniform|31.9|

So more than 60x speedup over the R version
on a four core laptop and about 30x for numpy.

> ## Work to do
> 
> Whilst we clearly have a significant performance boost
> over even the best-in-class Rust distribution system,
> we still have some work to do.
> 
> To prove the system, we need to do considerable verification
> work on the methods as well as generate R, python and GNU octave
> libraries.
> 
> There are many other functions which will benefit from the
> generalised system that **Doctor Syn** supports.
> 
> We should also look at super-accurate versions of these functions
> by using larger sizes, table lookups or fixed point integer
> arithmetic.
> 
> Work has started on support for ARM SVE, with work on the Isambard
> A64FX cluster underway.

Reviewer: You've got some great stuff here, I think this last paragraph should focus a bit more on telling everyone this.

```suggestion
## Work to do

With an implementation of just two Doctor Syn functions, we have shown a significant performance boost
over even the best-in-class Rust distribution system. This excellent result on a toy example shows excellent promise for what the generalised **Doctor Syn** system is capable of. 

Work has started on support for ARM SVE, with work on the (Isambard)[https://gw4.ac.uk/isambard/] A64FX supercomputing cluster underway. The Doctor Syn method also provides flexibility in the accuracy of the solution it provides, and we are exploring super-accurate versions of our functions by using larger sizes, table lookups or fixed point integer arithmetic. Finally, we are working on the substantial task of full verification as well as generation of R, python and Octave libraries. 

```
